### ISSUE TITLE
Disable constraints during data import

### ISSUE THREAD

**Issue by jimmymathews** [nadeemlab/SPT#61]
The current HALO import workflow is too slow, taking 35 minutes for a 2.5m cell dataset.
Here are a few facts of the matter:
- The `performance_report.md` shows that over 75% of the time is spent just doing the `COPY FROM` statement. That is, all the preparation of records, including shapefile creation, and memory mapping in mmap, are negligible at the moment, in comparison to the actual database insertion. More specifically it's about 12 seconds per upload of 100k records in one of the (~4) big tables.
- I did previously test the chunk size (for each `COPY FROM`) for effect on speed performance. The result was remarkably flat, in that total time was independent of chunk size (roughly 1k ~ 100k tested). In other words the culprit is probably not a per-transaction or per-commit overhead.
- The upload speed on the machine being tested is 1GB, so the bottleneck is definitely not internet speed.
- According to a little research, foreign key constraints and implicitly created indexes on primary keys are almost certainly the cause.

The widely-recommended corrective is to drop or disable the constraints before a bulk operation and re-enable them afterward.

There are many steps to accomplishing this, and there seem to be many ways to do it. The simplest, available in Postgres, may be of the form:

```
ALTER TABLE tablename DISABLE TRIGGER ALL;
<upload...>
ALTER TABLE tablename ENABLE TRIGGER ALL;
```

Doing it this way apparently fails to disable primary key constraints unless they were explicitly marked "deferrable" at table creation time. Thus we also need to alter the schema to include the DEFERRABLE keyword on the primary key declaration. Though this risks making the schema more SQL-vendor-specific, it's most likely worth it.

**Comment by jimmymathews** [comment]
Being developed as a branch off of issue54, which is where the most up-to-date HALO import functionality currently resides.

**Comment by jimmymathews** [comment]
Disabling triggers requires elevated database user privileges, and for good reason; data changes may break constraints in a way that is unnoticed. Thus an automated implementation, in which the disabling/enabling statements are added to the workflow, is a bit risky and also requires a second set of authentication credentials to be supplied, risking eventual bugs in which elevated privileges are used for other things.

Since bulk upload like this is meant to be infrequent and "managed", it may be safer to add a `spt-db-constraints enable` and `spt-db-constraints disable` command with its own `--database-config-file=...` argument, to be used to wrap pre and post `nextflow run .` in the run script `run.sh`, created on `spt-configure` in case `--database-config-file-elevated` is supplied.

**Comment by jimmymathews** [comment]
It seems that the foreign key constraints are also not deferrable by default, but can be made so at `CREATE TABLE ... REFERENCES ...` time.

**Comment by jimmymathews** [comment]
It turns out that `DEFERRABLE` primary keys cannot be referenced as foreign key targets. Because of this we may only be able to disable the foreign key constraints.

**Comment by jimmymathews** [comment]
The command has been implemented as of d7d2972ec29dd7d4b6ddf593577868710d0ab5c2.

Tested on a local db instance:

```
$ spt-db-constraints --database-config-file-elevated=~/.spt_db.config.local --disable

06-02 16:39:27 [  INFO   ] environment.database_connection: Found database credentials ['endpoint', 'database', 'user', 'password']
06-02 16:39:27 [  INFO   ] environment.database_connection: endpoint: localhost
06-02 16:39:27 [  INFO   ] environment.database_connection: database: pathstudies
06-02 16:39:27 [  INFO   ] environment.database_connection: user:     postgres
06-02 16:39:27 [  DEBUG  ] spt-db-constraints:76: Executing: ALTER TABLE histological_structure DISABLE TRIGGER ALL
06-02 16:39:27 [  DEBUG  ] spt-db-constraints:76: Executing: ALTER TABLE histological_structure_identification DISABLE TRIGGER ALL
06-02 16:39:27 [  DEBUG  ] spt-db-constraints:76: Executing: ALTER TABLE expression_quantification DISABLE TRIGGER ALL
06-02 16:39:27 [  DEBUG  ] spt-db-constraints:76: Executing: ALTER TABLE shape_file DISABLE TRIGGER ALL
06-02 16:39:27 [  INFO   ] spt-db-constraints: Showing trigger info:
06-02 16:39:27 [  DEBUG  ] spt-db-constraints:52: (for "enabled" code refer to "tgenabled" in https://www.postgresql.org/docs/current/catalog-pg-trigger.html)
enabled  deferrable  constraint_name                                                   relation_name                           
D        true        expression_quantification_histological_structure_fkey             expression_quantification               
D        true        expression_quantification_target_fkey                             expression_quantification               
D        true        expression_quantification_target_fkey                             expression_quantification               
D        true        expression_quantification_histological_structure_fkey             expression_quantification               
D        true        expression_quantification_histological_structure_fkey             histological_structure                  
D        true        histological_structure_identificati_histological_structure_fkey   histological_structure                  
D        true        histological_structure_identificati_histological_structure_fkey   histological_structure                  
D        true        expression_quantification_histological_structure_fkey             histological_structure                  
D        true        histological_structure_identificati_histological_structure_fkey   histological_structure_identification   
D        true        histological_structure_identification_data_source_fkey            histological_structure_identification   
D        true        histological_structure_identification_data_source_fkey            histological_structure_identification   
D        true        histological_structure_identification_shape_file_fkey             histological_structure_identification   
D        true        histological_structure_identification_shape_file_fkey             histological_structure_identification   
D        true        histological_structure_identif_plane_coordinates_reference_fkey   histological_structure_identification   
D        true        histological_structure_identif_plane_coordinates_reference_fkey   histological_structure_identification   
D        true        histological_structure_identificati_histological_structure_fkey   histological_structure_identification   
D        true        histological_structure_identification_shape_file_fkey             shape_file                              
D        true        histological_structure_identification_shape_file_fkey             shape_file
```

```
$ spt-db-constraints --database-config-file-elevated=~/.spt_db.config.local --enable

06-02 16:40:12 [  INFO   ] environment.database_connection: Found database credentials ['endpoint', 'database', 'user', 'password']
06-02 16:40:12 [  INFO   ] environment.database_connection: endpoint: localhost
06-02 16:40:12 [  INFO   ] environment.database_connection: database: pathstudies
06-02 16:40:12 [  INFO   ] environment.database_connection: user:     postgres
06-02 16:40:12 [  DEBUG  ] spt-db-constraints:76: Executing: ALTER TABLE histological_structure ENABLE TRIGGER ALL
06-02 16:40:12 [  DEBUG  ] spt-db-constraints:76: Executing: ALTER TABLE histological_structure_identification ENABLE TRIGGER ALL
06-02 16:40:12 [  DEBUG  ] spt-db-constraints:76: Executing: ALTER TABLE expression_quantification ENABLE TRIGGER ALL
06-02 16:40:12 [  DEBUG  ] spt-db-constraints:76: Executing: ALTER TABLE shape_file ENABLE TRIGGER ALL
06-02 16:40:12 [  INFO   ] spt-db-constraints: Showing trigger info:
06-02 16:40:12 [  DEBUG  ] spt-db-constraints:52: (for "enabled" code refer to "tgenabled" in https://www.postgresql.org/docs/current/catalog-pg-trigger.html)
enabled  deferrable  constraint_name                                                   relation_name                           
O        true        expression_quantification_target_fkey                             expression_quantification               
O        true        expression_quantification_histological_structure_fkey             expression_quantification               
O        true        expression_quantification_histological_structure_fkey             expression_quantification               
O        true        expression_quantification_target_fkey                             expression_quantification               
O        true        histological_structure_identificati_histological_structure_fkey   histological_structure                  
O        true        expression_quantification_histological_structure_fkey             histological_structure                  
O        true        expression_quantification_histological_structure_fkey             histological_structure                  
O        true        histological_structure_identificati_histological_structure_fkey   histological_structure                  
O        true        histological_structure_identif_plane_coordinates_reference_fkey   histological_structure_identification   
O        true        histological_structure_identificati_histological_structure_fkey   histological_structure_identification   
O        true        histological_structure_identification_data_source_fkey            histological_structure_identification   
O        true        histological_structure_identification_data_source_fkey            histological_structure_identification   
O        true        histological_structure_identification_shape_file_fkey             histological_structure_identification   
O        true        histological_structure_identification_shape_file_fkey             histological_structure_identification   
O        true        histological_structure_identificati_histological_structure_fkey   histological_structure_identification   
O        true        histological_structure_identif_plane_coordinates_reference_fkey   histological_structure_identification   
O        true        histological_structure_identification_shape_file_fkey             shape_file                              
O        true        histological_structure_identification_shape_file_fkey             shape_file
```

I took the opportunity to factor out connection-making from `skimmer.py`, in order to reuse it here. This connection credential finding and connecting is now done in `environment.database_connection.py`.

**Comment by jimmymathews** [comment]
d7d2972ec29dd7d4b6ddf593577868710d0ab5c2 is merged into issue54 branch #54 .

**Comment by jimmymathews** [comment]
This way of disabling constraints can not be implemented on RDS, due to limitations of the most-privileged user available.
It is slightly more work, but one can just drop and re-create the foreign key constraints instead.

**Comment by jimmymathews** [comment]
This is now implemented as `spt-db-constraints --drop` and `spt-db-constraints --recreate`.
A unit test was added that, in a local instance, initializes the db, drops the pertinent constraints, and recreates them.

Also tested on RDS with large/medium data set. Recreation of the 6 constraints took 1.5 minutes.

**Comment by jimmymathews** [comment]
Now, the effect of this constraint disabling/re-enabling on the dataset import. (Not RDS, different provider):
- Upload for 2.5 million cells: 10.5 minutes
- Recreation of constraints: 1 minute

Here is the performance/timing report:
```
| from                                         | to                                           |   average time spent |   total time spent |   frequency |    fraction |
|:---------------------------------------------|:---------------------------------------------|---------------------:|-------------------:|------------:|------------:|
| Created shapefile contents                   | Beginning of one cell iteration              |          9.01568e-05 |      231.937       |     2572591 | 0.395341    |
| Beginning of one cell iteration              | Created shapefile contents                   |          6.40419e-05 |      164.756       |     2572629 | 0.28083     |
| Started copy from command for bulk insertion | Finished inserting one chunk                 |          0.941074    |      143.043       |         152 | 0.24382     |
| Completed cell manifest parsing              | Retrieved and hashed a cell manifest         |          1.29245     |       27.1415      |          21 | 0.0462633   |
| Started encoding one chunk                   | Started inserting chunk into local memory    |          0.0754675   |       11.4711      |         152 | 0.0195526   |
| Started per-cell iteration                   | Beginning of one cell iteration              |          0.0899421   |        3.4178      |          38 | 0.00582571  |
| Starting a chunk                             | Subsetted cells dataframe on chunk           |          0.0582297   |        2.21273     |          38 | 0.00377164  |
| Started inserting chunk into local memory    | Started copy from command for bulk insertion |          0.00801009  |        1.21753     |         152 | 0.00207531  |
| Retrieved next integer identifiers           | Retrieved and hashed a cell manifest         |          0.764217    |        0.764217    |           1 | 0.00130262  |
| Subsetted cells dataframe on chunk           | Retrieved discretizations on chunk           |          0.0178569   |        0.678562    |          38 | 0.00115662  |
| Cursor opened                                | Retrieved next integer identifiers           |          0.0203917   |        0.0203917   |           1 | 3.4758e-05  |
| Initial                                      | Cursor opened                                |          0.00520447  |        0.00520447  |           1 | 8.87112e-06 |
| Finished inserting one chunk                 | Completed cell manifest parsing              |          0.000163895 |        0.00360569  |          22 | 6.14598e-06 |
| Retrieved discretizations on chunk           | Started per-cell iteration                   |          8.61841e-05 |        0.003275    |          38 | 5.5823e-06  |
| Created shapefile contents                   | Started encoding one chunk                   |          6.50728e-05 |        0.00247277  |          38 | 4.21488e-06 |
| Finished inserting one chunk                 | Started encoding one chunk                   |          4.7171e-06  |        0.000537749 |         114 | 9.16605e-07 |
| Retrieved and hashed a cell manifest         | Starting a chunk                             |          9.90165e-06 |        0.000217836 |          22 | 3.71306e-07 |
| Finished inserting one chunk                 | Starting a chunk                             |          1.07682e-05 |        0.000172291 |          16 | 2.93673e-07 |
```
Now over 75% of the time is spent on a parallelizable activity. The only difficulty here would be each parallel job predicting the correct integer identifiers for the cell-level data it processes. This is just a matter of allocation, no big deal.

The speed up should then leave us with about 4 to 5 minutes total time for data upload for 2.5m cell dataset. This will be another issue.

**Comment by jimmymathews** [comment]
Merged into issue54 as of 90bbde59a3ea8500f3ab13fa3bb0bff8e0e39fe1.
